src/Transport/Sending/MessageDispatcher.cs

```diff
@@ -96,17 +97,30 @@ public Task Dispatch(TransportOperations outgoingMessages, TransportTransaction
             return tasks.Count == 1 ? tasks[0] : Task.WhenAll(tasks);
         }
 
-        static void ApplyCustomizationToOutgoingNativeMessage(ReadOnlyContextBag context, IOutgoingTransportOperation transportOperation, Message message)
+        static void ApplyCustomizationToOutgoingNativeMessage(IOutgoingTransportOperation transportOperation,
+            Message message, TransportTransaction transportTransaction)
         {
-            if (transportOperation.Message.Headers.TryGetValue(CustomizeNativeMessageExtensions.CustomizationHeader, out var customizationId))
+            if (!transportOperation.Properties.TryGetValue(NativeMessageCustomizationBehavior.CustomizationKey,
+                out var key))
             {
-                if (context.TryGet<NativeMessageCustomizer>(out var nmc) && nmc.Customizations.Keys.Contains(customizationId))
-                {
-                    nmc.Customizations[customizationId].Invoke(message);
-                }
+                return;
+            }
+
+            if (!transportTransaction.TryGet(out NativeMessageCustomizer messageCustomizer))
+            {
+                Log.Warn(
```

Comment:
Is this possible? If the extension method to customize a native message is invoked, it will [always assign something to the `ExtendableOptions` or throw](https://github.com/Particular/NServiceBus.Transport.AzureServiceBus/pull/221/files#diff-59e67e37ca39f0c5562cfb976ea7dd1032bf644506503fb2aafb774e26a89c99R28).

Reply:
I think this might happen in Outbox scenarios if the dispatch fails and then is retried from the outbox storage as the operation properties might be persisted.

Reply:
this should really not happen so it might be fine to throw instead.